// @hammerhq/logger/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMockLogger = exports.createLogger = exports.generateDate = exports.icons = void 0;
const chalk_1 = require("chalk");
const log = (prefix, logs, icon, color) => console.log(generateDate(), color(icon), prefix, ...logs);
exports.icons = {
    infoIcon: chalk_1.supportsColor ? "ℹ" : "i",
    successIcon: chalk_1.supportsColor ? "✔" : "√",
    warningIcon: chalk_1.supportsColor ? "⚠" : "‼",
    errorIcon: chalk_1.supportsColor ? "✖" : "×",
    eventIcon: "☄",
    debugIcon: "☢",
    logIcon: "✎",
};
function generateDate() {
    const options = {
        hour: "numeric",
        minute: "numeric",
        second: "numeric",
        hour12: false,
    };
    const date = new Date().toLocaleString("en-US", options);
    const result = (0, chalk_1.grey)(date);
    return result;
}
exports.generateDate = generateDate;
function createLogger(prefix, icon, color) {
    return function logger(...logs) {
        log(prefix, logs, icon, color);
    };
}
exports.createLogger = createLogger;
const createMockLogger = (prefix, icon, color) => (...logs) => `${generateDate()} ${color(icon)} ${prefix} ${logs.join(", ")}`;
exports.createMockLogger = createMockLogger;
